### ISSUE TITLE
Pull request plugin for Che's Theia

### ISSUE THREAD

**Issue by ibuziuk** [redhat-developer/rh-che#778]
Need to create PR plugin for Che's Theia with minimal functionality and contribute it upstream. PR should be very simple and have minimal functionality / UI. After discussing this task with @evidolob it was decided to implement smth similar to VSCode `GitHub` plugin:

- User installs PR plugin (initially only GitHub will be supported)
- User generates personal access token on GitHub and set it via the command palette by executing dedicated command e.g. `PR: Set Personal Access Token for GitHub`. There also would be some notification regarding this:

![image](https://user-images.githubusercontent.com/1461122/43410592-ad742d56-9427-11e8-8e15-b3a5fd3d26a8.png)
- User  presses `GitHub` related button on the outline view that was added by the PR plugin

![image](https://user-images.githubusercontent.com/1461122/43412246-e918aad0-942c-11e8-9b2f-65bd7e8af216.png)

- Command pallete is opened and asks via `QuickOpenService` in which branch does the user want to send a PR
- Once user picks the branch the pop-up is shown where PR description should be provided. 
- Once PR description is provided and user presses OK button - PR is send and notification with the link to the PR is shown
- On the GitHub PR description would contain factory link to review the PR. Example: che-samples/web-java-spring-petclinic#31 (Once 'Ok' button is pressed from the previous step, under the hood the factory is created and  link to it is added to PR description)

NOTE:
in the first iteration it is planned that: 
- PRs would be always send from the current branch (no possibility to pick branch)
- PRs would be always send to the `origin` remote, or if there is only one remote exist  than in that one  (no possibility to pick remote) 

**Investigation tasks:**
- [ ] Get familiar with Theia plugin development - run / debug / update basic `hello-world` plugin #786 
- [ ] Create a POC of PR plugin (UI only without backend) #788
- [ ] Make research about available node.js libraries which provide GitHub integration in order to choose the right one for PR plugin (lib must be EPL 2.0 compatible) #789
- [ ] Investigate a possibility to reuse https://github.com/eclipse/che-theia-github-plugin

**Comment by slemeur** [comment]
Sounds good @ibuziuk ! 

Just one comment, in the PR we are setting up, we need the user to provide a description and we will also put the factory link to review the PR. 
Example: https://github.com/che-samples/web-java-spring-petclinic/pull/31

**Comment by ibuziuk** [comment]
@slemeur thanks for review. Description & factory link have been added to this epic

**Comment by slemeur** [comment]
BTW, I think we already have some GitHub extension: https://github.com/eclipse/che-theia-github-plugin

**Comment by ibuziuk** [comment]
@slemeur hmm.. I was not aware of this extension
@benoitf @evidolob could you please provide some info about it (maybe there is some demo) ? how different is it from the plugin description from this issue ? Should we start from scratch or it would be better to use  https://github.com/eclipse/che-theia-github-plugin as a basement e.g. using the same UI / GitHub lib etc.

**Comment by ibuziuk** [comment]
@evidolob ok, thanks for the info - we would probably need to do some investigation first.
@benoitf do you happen to know if there are any docs / instructions how to add this plugin to Theia (Readme does not contain any relevant info) ?

**Comment by ibuziuk** [comment]
@evidolob could you please take a look at the investigation tasks I have added to the description which we are planning to add to the next sprint? maybe you have any suggestions? Also in terms of docs I believe we could use in order to start: 

- https://github.com/theia-ide/theia/blob/master/doc/Developing.md
- https://www.theia-ide.org/doc/index.html

**Comment by ibuziuk** [comment]
@evidolob plugin of course ;-)

**Comment by evidolob** [comment]
https://github.com/eclipse/che-theia-github-plugin is a Theia extension :) 
In case that you want to write plugin, you cannot use `che-theia-github-plugin`, you can use `@octokit/rest` library.
AFAIK `che-theia-github-plugin` was created before we start work on plugins, so yes name is confusing

**Comment by slemeur** [comment]
would it be relevant to keep the `che-theia-github-plugin` then? I think the scope describe in this issue is covering the capabilities available in `che-theia-github-plugin`, or should it combine its current capabilities with the ones list in the issue?

**Comment by evidolob** [comment]
I guess not, I like idea to have one Theia plugin with all GitHub related functionality

**Comment by slemeur** [comment]
yes I agree :)

**Comment by ibuziuk** [comment]
@slemeur What we were planning to do initially  is to write new PR plugin from scratch using new Theia plugin model and than existing GitHub extension [1] was mentioned in the issue which introduced some confusion. 

[1] https://github.com/eclipse/che-theia-github-plugin

**Comment by benoitf** [comment]
Documentation on Theia plug-ins is "in progress", PR was made there https://github.com/theia-ide/theia-website/pull/10

**Comment by ibuziuk** [comment]
@slemeur @l0rd We are planning to take the following issues to the sprint. Please, comment if those look good:

- [ ] Get familiar with Theia plugin development - run / debug / update basic `hello-world` plugin #786 
- [ ] Create a POC of PR plugin (UI only without backend) #788
- [ ] Make research about available node.js libraries which provide GitHub integration in order to choose the right one for PR plugin (lib must be EPL 2.0 compatible) #789

**Comment by slemeur** [comment]
That sounds a reasonable approach.

**Comment by ibuziuk** [comment]
thanks for the input - issues are in the sprint

**Comment by ibuziuk** [comment]
Closing as outdated
